datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachService.java

```diff
@@ -0,0 +1,132 @@
+                reattachExternalDatabase(reattached, detachedCrn);
+            }
+
+            return reattached;
+        });
+    }
+
+    /**
+     * Throws RuntimeException if cluster is not detached or could not revert name and CRN.
+     */
+    public SdxCluster reattachCluster(SdxCluster cluster) {
+        LOGGER.info("Started reattaching SDX cluster with ID: {}", cluster.getId());
+
+        if (!cluster.isDetached()) {
+            throw new RuntimeException("Attempting to reattach a cluster which was not detached!");
+        }
+
+        sdxAttachDetachUtils.updateClusterNameAndCrn(
+                cluster, sdxDetachNameGenerator.generateOriginalNameFromDetached(cluster.getClusterName()),
+                cluster.getOriginalCrn()
+        );
+        cluster.setOriginalCrn(null);
+        cluster.setDetached(false);
+        SdxCluster saved = sdxClusterRepository.save(cluster);
+
+        LOGGER.info("Finished reattaching SDX cluster with ID: {}. Cluster now has name {} and crn {}.",
+                saved.getId(), saved.getClusterName(), saved.getCrn());
+        return saved;
+    }
+
+    /**
+     * Can throw the following exceptions:
+     *      - NotFoundException : Internal stack is not found by the cluster's name.
+     *      - TransactionExecutionException : If updating the stack fails.
+     */
+    public void reattachStack(SdxCluster cluster, String originalName) {
+        LOGGER.info("Started reattaching stack of SDX cluster with ID: {}", cluster.getId());
+        sdxAttachDetachUtils.updateStack(cluster, originalName);
+        LOGGER.info("Finished reattaching stack of SDX cluster with ID: {}. Now has name {} and crn {}.",
+                cluster.getId(), cluster.getClusterName(), cluster.getCrn());
+    }
+
+    /**
+     * Throws a NotFoundException if the database is not found for the cluster CRN and environment CRN.
+     */
+    public void reattachExternalDatabase(SdxCluster cluster, String detachedCrn) {
+        LOGGER.info("Started reattaching external database for SDX cluster with ID: {}", cluster.getId());
+        sdxAttachDetachUtils.updateExternalDatabase(cluster, detachedCrn);
+        LOGGER.info("Finished reattaching external database for SDX cluster with ID: {}. Now has crn: {}.",
+                cluster.getId(), cluster.getCrn());
+    }
+
+    /**
+     * Note that this method returns a possibly different version of the original cluster, so the original should
+     * not be used anymore.
+     *
+     * Will throw a TransactionExecutionException if:
+     *      - Cluster cannot be saved.
+     *      - The resource owner role assignment fails its UMS GRPC call.
+     */
+    public SdxCluster saveSdxAndAssignResourceOwnerRole(SdxCluster cluster) throws Exception {
+        return transactionService.required(() -> {
+            SdxCluster created = sdxClusterRepository.save(cluster);
+            ownerAssignmentService.assignResourceOwnerRoleIfEntitled(
```

Comment:
external calls shouldn't be part of a transaction

Reply:
ok, thanks. let's wait for @kkalvagadda1 approval then

Reply:
The majority of the places in the code, transaction service is used to wrap calls to repositories so that changes to the repository can be reverted when there are exceptions to achieve automicity.

Reply:
I did not realize that assignResourceOwnerRoleIfEntitled makes a call to UMS service.
I'm good with this change. Let's access to update the owner assignment when resize is done. we could maintain the older owner assignment.

let's consider another PR for this.